Add pubsub module for rpc info subscriptions

CriesofCarrots opened the pull request:
Fixes #1339 

This module adds a `subscriptionChannel` method to the JSON-RPC service with opens a websocket for subscription requests.
At present, this function will open whatever port the client requests, if it is free. This should be limited in some fashion, but I'm wondering about the best approach. Should the node select an appropriate port without client input? Or allow it to be customizable within our range?

mvines commented on doc/json-rpc.md:
```diff
@@ -227,3 +234,115 @@ curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc":"2.0","id":1, "m
+Open a socket on the node for JSON-RPC subscription requests
+
+##### Parameters:
+* `integer` - port to open for websocket subscription requests
```
Why does the caller suggest a port?  It seems like it would be better if the node could choose the port to use and return it

CriesofCarrots replied:
Sure, sounds like you prefer the first approach I mention above :)

mvines commented on doc/json-rpc.md:
```diff
@@ -227,3 +234,115 @@ curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc":"2.0","id":1, "m
+* `integer` - port to open for websocket subscription requests
+
+##### Results:
+* `string` - "addr", open websocket address
```
I'd just return the port here and not the full address that the node thinks it has (there might be a nat in between..)

mvines commented on doc/json-rpc.md:
```diff
@@ -23,6 +23,13 @@ Methods
 * [sendTransaction](#sendtransaction)
+* [subscriptionChannel](#subscriptionChannel)
+
+* [Subscription Websocket](#Subscription Websocket)
```
this link in the rendered doc doesn't work

mvines commented on doc/json-rpc.md:
```diff
@@ -227,3 +234,115 @@ curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc":"2.0","id":1, "m
+---
+
+### accountSubscribe
+*alias: accountSub*
```
IMO let's drop all the aliases, seems like it just adds confusion (ie, an option...)

CriesofCarrots replied:
👌

mvines commented on doc/json-rpc.md:
```diff
@@ -227,3 +234,115 @@ curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc":"2.0","id":1, "m
+
+---
+
+### Subscription Websocket
```
I don't see the message format documented when a subscribed event occurs?

mvines commented on doc/json-rpc.md:
```diff
@@ -227,3 +234,115 @@ curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc":"2.0","id":1, "m
+
+### Subscription Websocket
+After opening a subscription socket with the `subscriptionChannel` JSON-RPC request method, submit subscription requests via websocket protocol
+Connect to the websocket at `ws://<ADDRESS>/<PATH>` returned from the request
```
Would it be more convenient to send the subscribe/unsubscribe events on the normal RPC channel, then use the websockets for push only?   I suspect this might be easier for the client

mvines replied:
hmm, let's do what's convenient for the module side for now and revisit later if it's too annoying from the client side

mvines commented on doc/json-rpc.md:
```diff
@@ -227,3 +234,115 @@ curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc":"2.0","id":1, "m
+Connect to the websocket at `ws://<ADDRESS>/<PATH>` returned from the request
+- Submit subscription requests to the websocket using the methods below
+- Multiple subscriptions may be active at once
+- The subscription-channel socket will close when client closes websocket. To create new subscriptions, send a new `subscriptionChannel` JSON-RPC reqest.
```
"reqest"

mvines commented on src/pubsub.rs:
```diff
@@ -0,0 +1,614 @@
+                                .unwrap();
+                            break;
+                        }
+                        sleep(Duration::from_millis(100));
```
Can we plumb this directly into the bank perhaps to avoid the polling?

mvines commented on src/pubsub.rs:
```diff
@@ -0,0 +1,614 @@
+                        previous_userdata = Some(account.userdata);
+                    }
+                }
+                sleep(Duration::from_millis(100));
```
Oh, we definitely want to plug into the bank for account userdata changes somehow.

CriesofCarrots commented:
@mvines , could you take a look at how I implemented the subscription notifications inside the bank?
I added the WIP label because technically the unsubscribe methods are broken. They work on the bank end (as per tests), but I need to update the solana-jsonrpc-pubsub crate to pass along metadata in unsubscribe methods in order to access the `request_processor` (ie. bank) for removal.

In the meantime, it would be helpful to get your feedback on whether I'm on track. Thank you!

mvines commented on src/bank.rs:
```diff
@@ -936,6 +959,54 @@ impl Bank {
+        subscriptions.insert(pubkey, sink);
+    }
+
+    pub fn remove_account_subscrtion(&self, pubkey: Pubkey) -> bool {
```
nit: remove_account_subscrtion

mvines commented on src/bank.rs:
```diff
@@ -548,6 +565,12 @@ impl Bank {
         }
+        // Check account subscriptions
+        for ((pre, pubkey), post) in pre_userdata.iter().zip(program_accounts.iter()) {
+            if pre != &post.userdata {
```
I think it should be faster to first check if the account pubkey has been subscribed to, then memcmp the account for change second.  This way mem compares each account at the end of each transaction right?

mvines commented on src/bank.rs:
```diff
@@ -936,6 +959,54 @@ impl Bank {
+
+    fn check_signature_subscriptions(&self, signature: &Signature) {
+        if let Some(sink) = self.signature_subscriptions.write().unwrap().get(signature) {
+            sink.notify(Ok(RpcSignatureStatus::Confirmed))
```
Automatically unsubscribing a signature subscription on confirmation seems nice for the client, as there's only ever a single notification for a given signature.

Do we notify on failure too (such as ProgramFailure)?

CriesofCarrots replied:
Nice, I like unsubscribe on confirmation.
We don't currently notify on failure, but I can piece that in. You want it?

mvines replied:
Yes please.  Let's surface all final errors (ie, everything but `SignatureNotFound`), to help the client move along faster than waiting for an eventual timeout due to no successful confirmation.

CriesofCarrots commented:
@mvines ready for your eyes!

mvines commented on src/bank.rs:
```diff
@@ -942,23 +969,79 @@ impl Bank {
+        }
+    }
+
+    pub fn store_signature_subscription(
```
The name `store_signature_subscription` confused me for a moment.  Perhaps `add_signature_subscription` (same comment for `store_account_subscription`)

mvines commented on src/bank.rs:
```diff
@@ -498,6 +516,11 @@ impl Bank {
             .collect();
+        let pre_userdata: Vec<_> = program_accounts
+            .iter_mut()
+            .map(|a| a.userdata.clone())
```
I'm a little unsettled by this `userdata.clone()`, I feel like it'll hurt performance a little as this occurs on every transaction on every node even if are no account subscriptions.   But I have no data to back that up concretely.  Is there a quick way we can check if the node has at least one active account subscriptions, and if not skip all the cloning?    That's about the max I'd want to optimize here before actually measuring

mvines commented on src/bank.rs:
```diff
@@ -942,23 +969,79 @@ impl Bank {
+    }
+
+    fn check_account_subscriptions(&self, pubkey: &Pubkey, pre_userdata: &[u8], account: &Account) {
+        for (_bank_sub_id, (id, sink)) in self.account_subscriptions.read().unwrap().iter() {
```
Little sucky that this will be `O(N)` with the number of active subscriptions (`check_signature_subscriptions` has the same runtime looks like).  

I'm ok with this for now but please file a follow-up performance issue for this so we can remember to maybe revisit later.  We can roll in potentially optimizing the `a.userdata.clone()`s in this issue as well.

mvines commented on src/pubsub.rs:
```diff
@@ -0,0 +1,635 @@
+}
+
+pub struct PubSubService {
+    _thread_hdl: JoinHandle<()>,
```
why the `_`?  `_thread_hdl` seems used?

CriesofCarrots replied:
I think this must be because I don't use the handle in `rpc.rs`
The compiler complains if I remove the underscore.

mvines commented on src/pubsub.rs:
```diff
@@ -0,0 +1,635 @@
+            .name("solana-pubsub".to_string())
+            .spawn(move || {
+                let mut io = PubSubHandler::default();
+            	let rpc = RpcSolPubSubImpl::default();
```
`cargo fmt` allowed this?

CriesofCarrots replied:
That's odd. Those indentations look correct and aligned for me locally, and git says everything is up to date with the remote.
That said, `cargo fmt` seems to be ignoring the `Builder` expression when I deliberately screw it up.
I'll try formatting it from scratch and push with some other changes and see if that makes a difference.

mvines commented on src/pubsub.rs:
```diff
@@ -0,0 +1,635 @@
+                }
+                loop {
+                    if exit.load(Ordering::Relaxed) {
+                        for (_, bank_sub_id) in subs.read().unwrap().iter() {
```
lines 79-84 and lines 88-93 are duplicated, let's undup them

CriesofCarrots commented on doc/json-rpc.md:
````diff
@@ -227,3 +234,121 @@ curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc":"2.0","id":1, "m
+
+##### Notification Format:
+```bash
+{"jsonrpc": "2.0","method": "signature_notification", "params": {"result": "Confirmed","subscription":0}}
````
Ssssneaky

mvines commented on src/bank.rs:
```diff
@@ -276,7 +276,9 @@ impl Bank {
                 Err(_) => RpcSignatureStatus::GenericFailure,
             };
-            self.check_signature_subscriptions(&tx.signature, status);
+            if status != RpcSignatureStatus::SignatureNotFound {
```
short circuit even sooner?

```rust
if res[i] != Err(BankError::SignatureNotFound) {
     self.check_signature_subscriptions(&tx.signature,
         match res[i] {
                Ok(_) => RpcSignatureStatus::Confirmed,
                Err(BankError::ProgramRuntimeError(_)) => RpcSignatureStatus::ProgramRuntimeError,
                Err(BankError::SignatureNotFound) => RpcSignatureStatus::SignatureNotFound,
                Err(_) => RpcSignatureStatus::GenericFailure,
        }
   );
}
```
